addons/crm_addons/controllers/main.py

```diff
@@ -0,0 +1,566 @@
+        b64_encoded_signature = base64.b64encode(signature.encode()).decode()
+        return json.dumps({
+            "access_token": b64_encoded_string + "." + b64_encoded_signature,
+            "expires_in": timedelta(days=30).total_seconds(), # For gmail.
+        })
+
+    def _iap_enrich(self, domain):
+        #normalized_email = tools.email_normalize(partner.email)
+        #if not normalized_email:
+        #    return {}
+        #emails = {0: normalized_email.split('@')[1]}
+        domains = {0: domain}
+
+        enriched_data = {}
+        #enriched_data["enrichment_error"] = {'type': 'none', 'info': ''}
+        try:
+            response = request.env['iap.enrich.api'].with_user(request.env.user)._request_enrich(domains)
+            enriched_data = response.get("0")
+        except odoo.addons.iap.models.iap.InsufficientCreditError as ice:
+            _logger.info('Enrich request: failed because of credit')
+            enriched_data['enrichment_error'] = {'type': 'insufficient_credit', 'info': request.env['iap.account'].with_user(request.env.user).get_credits_url('reveal')}
+        except Exception as e:
+            enriched_data["enrichment_error"] = {'type': 'other', 'info': 'Unknown reason'}
+        else:
+            _logger.info('Enrich request: success')
+            if enriched_data == False:
+                enriched_data = {'enrichment_error': {'type': 'no_data', 'info': 'The enrichment API found no data for the email provided.'}}
+        return enriched_data
+
+    def _iap_enrich_partner_from_response(self, env, iap_data, partner):
+        enriched = {}
+        if not partner.name and iap_data.get("name"):
+            enriched["name"] = iap_data.get("name")
+        if not partner.street and iap_data.get("location"):
+            enriched["street"] = iap_data.get("location")
+        if not partner.city and iap_data.get("city"):
+            enriched["city"] = iap_data.get("city")
+        if not partner.zip and iap_data.get("postal_code"):
+            enriched["zip"] = iap_data.get("postal_code")
+        if not partner.phone and iap_data.get('phone_numbers'):
+            enriched['phone'] = iap_data['phone_numbers'][0]
+        if not partner.website and iap_data.get("domain"):
+            enriched['website'] = iap_data['domain']
+        country = partner.country_id
+        if not partner.country_id and iap_data.get('country_code'):
+            country = env['res.country'].search([('code', '=', iap_data['country_code'].upper())])
+            enriched['country_id'] = country.id
+        if not partner.state_id and country and iap_data.get('state_code'):
+            state = env['res.country.state'].search([
+                ('code', '=', iap_data['state_code']),
+                ('country_id', '=', country.id)
+            ])
+            enriched['state_id'] = state.id
+
+        partner.iap_enrich_info = iap_data
```

Comment:
Not data displayed in the add-on are storable in the current fields of Partner. Ex: "Industry", "Employees", "Year Founded", etc. Moreover we decided to store everything we could in case we want to display even more data in the future.